CHANGELOG.md

```diff
@@ -25,6 +25,35 @@
   queries don't return the same results as before, if they query for tags as part of the metric
   name.
   #1619
+* as of v0.13.1-250-g21d1dcd1 metrictank no longer excessively aligns all data to the same
+  lowest comon multiple resolution, but rather keeps data at their native resolution when possible.
+  1. When queries request mixed resolution data, this will now typically result in larger response datasets,
+  with more points, and thus slower responses.
+  Though the max-points-per-req-soft and max-points-per-req-hard settings will still help curb this problem.
+  Note that the hard limit was previously not always applied correctly.
+  Queries may run into this limit (and error) when they did not before.
+  2. This version introduces 2 new optimizations (see pre-normalization and mdp-optimization settings).
+  The latter is experimental and disabled by default, but the former is recommended and enabled by default.
+  It helps with alleviating the extra cost of queries in certain cases
+  (See https://github.com/grafana/metrictank/blob/master/docs/render-path.md#pre-normalization for more details)
+  When upgrading a cluster in which you want to enable pre-normalization (recommended),
+  you must apply caution: pre-normalization requires a PNGroup property to be
+  communicated in intra-cluster data requests, which older peers don't have.
+  The peer receiving the client request, which fans out the query across the cluster, will only set
+  the flag if the optimization is enabled (and applicable).  If the flag is set for the requests,
+  it will need the same flag set in the responses it receives from its peers in order to tie the data back to the initiating requests. 
+  Otherwise, the data won't be included in the response, which may result in missing series, incorrect aggregates, etc.
+  Peers responding to a getdata request will include the field in the response, whether it has the
+  optimization enabled or not.
+
+  Thus, to upgrade an existing cluster, you have 2 options:
+  A) disable pre-normalization, do an in-place upgrade. enable it, do another in-place upgrade.
+     This works regardless of whether you have a separate query peers, and regardless of whether you first
+     upgrade query or shard nodes.
+  B) do a colored deployment: create a new gossip cluster that has the optimization enabled from the get-go,
+     then delete the older deployment.
+
+  Note that you should not upgrade to to v0.13.1-250-g21d1dcd1, but to v0.13.1-252-ge2fa083 or later due to a critical bugfix.
```

Comment:
```suggestion
  Note that you should not upgrade to to v0.13.1-250-g21d1dcd1, but to v0.13.1-266-g80ca54f or later due to critical bug fixes.
```
There were 1-2 more bug fixes related to the align requests changes. The shadowed var fix was the latest one and I haven't found any more yet.

Reply:
so what do you suggest. we can't possibly always mention all future bugfixes relating to the change we're describing, nor should we. I already added this line to try to accommodate you but frankly we should probably just switch to how it was. One should always operate on the assumption that bugfixes may come after a feature is introduced. If people want as-bugfree-as-possible code they should wait until we tag a stable release.

Reply:
I agree, but since the fixes are in before this was merged we should probably update it to reflect the latest known version with the bug fixes, or as you suggested just remove it.

Reply:
no, a changelog can never accurately describe the current state of affairs, and trying to make it that for this particular case sets the wrong expectations. i just removed this again.